code/controllers/subsystems/fluids.dm

```diff
@@ -41,14 +39,36 @@ var/datum/controller/subsystem/fluids/SSfluids
 		active_fluids_copied_yet = FALSE
 		af_index = 1
 
-	var/flooded_a_neighbor // Not used, required by FLOOD_TURF_NEIGHBORS.
 	var/list/curr_sources = processing_sources
+	var/list/checked = list()
 	while (curr_sources.len)
 		var/turf/T = curr_sources[curr_sources.len]
 		curr_sources.len--
-
-		FLOOD_TURF_NEIGHBORS(T, FALSE)
-
+		var/flooded_a_neighbor
+		for(var/spread_dir in GLOB.cardinal)
+			UPDATE_FLUID_BLOCKED_DIRS(T)
+			if(T.fluid_blocked_dirs & spread_dir) 
+				continue
+			var/turf/next = get_step(T, spread_dir)
+			if(!istype(next) || next.flooded) 
+				continue
+			UPDATE_FLUID_BLOCKED_DIRS(next)
+			if((next.fluid_blocked_dirs & GLOB.reverse_dir[spread_dir]) || !next.CanFluidPass(spread_dir) || checked[next])
+				continue
+			checked[next] = TRUE
+			flooded_a_neighbor = TRUE
+			var/obj/effect/fluid/F = locate() in next
+			if(!F)
+				F = new /obj/effect/fluid(next)
+				var/datum/gas_mixture/GM = T:return_air()
+				if(GM) F.temperature = GM.temperature
+			if(F)
+				if(F.fluid_amount < FLUID_MAX_DEPTH)
+					SET_FLUID_DEPTH(F, FLUID_MAX_DEPTH)
+		if(!flooded_a_neighbor)
+			REMOVE_ACTIVE_FLUID_SOURCE(T)
+		else
+			ADD_ACTIVE_FLUID_SOURCE(T)
```

Comment:
Isn't this redundant; i.e. if we've reached this line, aren't we in the relevant list already?